### DISCUSSION TITLE
Include output of "longer" external process into the TUI without delay (missing async/thread knowledge)

### DISCUSSION THREAD

**Discussion by lukeflo** [ratatui/ratatui#1554] [Q&A] [answered]
Hi,
just for fun I write a little TUI which should work as kind of a wrapper for my package manager. Since the pm itself offers no usable API for Rust, I run most commands as external child processes using `std::process::Command`.

Generally it works OK, but I ran into a problem with rendering. The app is divided into two main parts. On the left a list of all packages and on the right the info for the currently selected package. To render the info I need to run the particular external command which needs about 0.3 to 0.5s to complete. This leads to a delay if you're scrolling the list very fast. I tried some approaches to run the command parallel in another thread but couldn't figure it out. The problem I'm facing is likely bound to the fact that the `terminal.draw` call redraws the TUI every tick/keyevent and thus always respawns the command to get the package info. I do not know how to spawn the process outside of the draw loop but include the output as soon as the program finished the run. I'm using the ratatui simple async template as foundation, so async tasks should be no problem in general. But I'm not too experienced with those Rust thread stuff.

I am thankful for any suggestion how to accomplish that.

I'll also try to provide some code examples asap, but right now I write from my smartphone so I can't access my local repo.

**Comment by joshka** [comment, marked as answer]
Your best bet is to start by reading tokio tutorial. It gives a good foundation for being able to understand the answers for how to do most things in async rust.

From there, the general solution is some form of shared state. There's a few main types of this:
- shared memory access, where both tasks hold a copy of an `Arc<Mutex<T>>`
- channels, tasks send messages when they're done using e.g. `mpsc::channel()` or `oneshot::channel()`
- atomic variables (e.g. AtomicBool, etc.)

If you care about seeing all the output added to the tui when it comes in (regardless if the user has navigated away), then you probably should use either a `Arc<Mutex>` wrapping some struct that stores a list of results, or an mpsc::channel, which gets a message on each line of the command output.

If you only really care about the one return value of the command, I'd use a oneshot. If you don't care about the result of a command if the user has navigated away, you can drop the receiver side when the selection changes and the output will just be ignored.

For the arc mutex approaches, just lock the state when reading / writing. Wrap the `Arc<Mutex>` in a struct to make this easy to encapsulate the locking / reading / writing. Pass a cloned copy of this to each task you create. Grab a copy of the output before rendering.

For the channels approach, pass the sender side of the channel to each task, do a try_read before you render the frame.

The biggest thing to avoid is locking the tasks for too long, because then you have other problems (see https://ryhl.io/blog/async-what-is-blocking/)

---

If you're fairly new to rust, I'd suggest avoiding async for a bit and keeping with normal threads. The techniques are pretty similar (swapping tokio types for std::sync types mostly and tasks for threads.

---

I'd also recommend doing a bit of a search on the discord / forums for async info. There's a bunch of useful info that's been added over time there. Someday we'll summarize it.

**Reply by lukeflo** [reply]
Thank you for the comprehensive answer. I'm using Rust since half a year now, don't know if that's totally new, but definitely haven't touched many advanced techniques 😉 

> If you only really care about the one return value of the command, I'd use a oneshot. If you don't care about the result of a command if the user has navigated away, you can drop the receiver side when the selection changes and the output will just be ignored.

That sounds like the best solution. If I scrolling the list fast, I don't care about the output because there is no chance to read it.

Will have a look on the resources you mentioned. As always, there seems to be no abbreviation to understand those things... 😁 

Thanks for your help. Keep up the great work!

**Reply by lukeflo** [reply]
Just in case someone is having a similar problem. I found a good working solution using some `Arc`, `Mutex` and `Condvar` stuff: [Solution with code example](https://stackoverflow.com/a/79298245/19647155)

**Comment by lukeflo** [comment]
I tried to get it working just with `std` lib stuff. But I only have a semi-good solution which doesn't block the main thread but runs into problems when scrolling the list very fast.

To not spam this forum I've opened a question on [stackoverflow](https://stackoverflow.com/q/79275324/19647155) with MWE. Just in case someone is looking for a similar problem...
